app/gui2/src/composables/stackNavigator.ts

```diff
@@ -0,0 +1,118 @@
+import type { BreadcrumbItem } from '@/components/NavBreadcrumbs.vue'
+import { useGraphStore } from '@/stores/graph'
+import { useProjectStore } from '@/stores/project'
+import { qnLastSegment, tryQualifiedName } from '@/util/qualifiedName.ts'
+import { computed, onMounted, ref } from 'vue'
+import type { StackItem } from '../../shared/languageServerTypes.ts'
+import type { ExprId } from '../../shared/yjsModel.ts'
+
+export function useStackNavigator() {
+  const projectStore = useProjectStore()
+  const graphStore = useGraphStore()
+
+  const breadcrumbs = ref<StackItem[]>([])
+
+  const breadcrumbLabels = computed(() => {
+    const activeStackLength = projectStore.executionContext.desiredStack.length
+    return breadcrumbs.value.map((item, index) => {
+      const label = stackItemToLabel(item)
+      const isActive = index < activeStackLength
+      return { label, active: isActive } as BreadcrumbItem
+    })
+  })
+
+  const allowNavigationLeft = computed(() => {
+    return projectStore.executionContext.desiredStack.length > 1
+  })
+
+  const allowNavigationRight = computed(() => {
+    return projectStore.executionContext.desiredStack.length < breadcrumbs.value.length
+  })
+
+  function stackItemToLabel(item: StackItem): string {
+    switch (item.type) {
+      case 'ExplicitCall':
+        return item.methodPointer.name
+      case 'LocalCall': {
+        const exprId = item.expressionId
+        const info = graphStore.db.getExpressionInfo(exprId)
+        return info?.methodCall?.methodPointer.name ?? 'unknown'
+      }
+    }
+  }
+
+  function handleBreadcrumbClick(index: number) {
+    const activeStack = projectStore.executionContext.desiredStack
+    if (index < activeStack.length) {
+      const diff = activeStack.length - index - 1
+      for (let i = 0; i < diff; i++) {
+        projectStore.executionContext.pop()
+      }
+    } else if (index >= activeStack.length) {
+      const diff = index - activeStack.length + 1
+      for (let i = 0; i < diff; i++) {
+        const stackItem = breadcrumbs.value[index - i]
+        if (stackItem?.type === 'LocalCall') {
+          const exprId = stackItem.expressionId
+          projectStore.executionContext.push(exprId)
+        } else {
+          console.warn('Cannot enter non-local call.')
+        }
+      }
+    }
+    graphStore.updateState()
+  }
+
+  function enterNode(id: ExprId) {
+    const expressionInfo = graphStore.db.getExpressionInfo(id)
+    if (expressionInfo == undefined || expressionInfo.methodCall == undefined) {
```

Comment:
prefer `x == null` (or `!x` if we want to ignore all falsy values) over `x == undefined`